Galran/new geometry module

galran opened the pull request:
Introducing the new Geometry module, which currently contains types (Vec3, Vec4 and Transform) and some utility functions.
The Transform type is a replacement for the RigidBodyTransform that was part of the CSG part of the package. The new Transform is used to position surfaces inside a CSG tree, but also emitters in 3D and everywhere else a 3D transform is needed.

The changes are all over the package, including the documentation, so please take the time and go through them.

alfredclwong reviewed (commented):
Looks great, just a few minor comments.

galran commented on docs/src/basic_types.md:
```diff
@@ -0,0 +1,58 @@
+# Basic Types
+
+The following are basic geometric types and utilities, such as vectors and transforms, that are used all acoross the package. Using these types requare, in addition to the `using OpticSim` statement to also use the OpticSim.Geoemtry module like:
```
fixed

galran commented on src/Geometry/Geometry.jl:
```diff
@@ -20,6 +20,9 @@
 # SOFTWARE
 
+include("Transform.jl")
+using .Geometry
```
fixed

galran commented on src/Geometry/CSG/Intersection.jl:
```diff
@@ -183,3 +183,18 @@ israyorigin(::RayOrigin) = true
+    i = interface(intsct)
+    if VERSION < v"1.6.0-DEV"
+        # TODO REMOVE
+        return @unionsplit OpticalInterface T i Intersection(α(intsct), a * point(intsct), Geometry.rotate(a, normal(intsct)), u, v, i, flippednormal = flippednormal(intsct))
```
@BrianGun do you know what this is?

BrianGun replied:
In 1.5 if you had unions of concrete types with more than 4 (or maybe more than 3?) types then the compiler wouldn't optimize the code properly and would allocate on the heap. 

1.6 should allow unions of greater than 4 but I don't think we've tested this yet. If you get rid of it then performance on 1.5 for multithreaded code could be terrible, as in slower than single threaded. We have to decide how much we care about continuing to support 1.5 since 1.6 is so much better.

alfredclwong commented:
Do we think this merits a bump to v0.4.0 (breaking change)?

BrianGun reviewed (commented):
@galran you don't need to define zero3 and one3. zero(Vec3{Float64}), ones(Vec3{Float64}) will work, as will rand(Vec3{Float64}). Same for Vec4. 

Not sure it's necessary to define origin as well as zero but it's simpler to define it this way because you only have to define it once for any dimension of Vec:

 origin(::Type{S}) where{S<:FieldVector} = zero(S)

the constructors for Vec3 and Vec4 for zero entries could also be simplified from this:

function Vec4(::Type{T} = Float64) where {T<:Real}
    return Vec4{T}(zero(T), zero(T), zero(T), zero(T))
end

to this

 Vec4(::Type{T}) where{T<:Real} = zero(Vec4{T})

instead of creating the identityT functions you can use 

I (capital i) 

wherever you need an identity matrix instead. You don't need to define the identityT functions. I is a generic identity matrix which the linear algebra packages know how to operate with but don't instantiate. https://stackoverflow.com/questions/57270276/identity-matrix-in-julia has a concise discussion.

for extracting the translation and rotation components of a Transform you can use the @view macro:

rotation(a::Transform) = @view a[1:3,1:3]

I'm pretty sure that in 1.6 @view is allocation free. 

You define struct Transform{T} <: FieldMatrix{4, 4, T}
    _xx::T
    _yx::T
    _zx::T
    _wx::T
    _xy::T
    _yy::T
    _zy::T
    _wy::T
    _xz::T
    _yz::T
    _zz::T
    _wz::T
    _xw::T
    _yw::T
    _zw::T
    _ww::T
end
export Transform

but then nothing seems to use these field names. Might be simpler to define Transform this way:

Transform{T} = SMatrix{4,4,T,16}

which you would instantiate like this

 Transform{Float64}(1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16)

galran commented:
I fixed everything except the 'I' - i don't understand exactly if this is a "laze" definition or how to use it efficiently, so i left the identitytransform function for now (i removed the identityT function).

BrianGun commented:
@galran did you push the latest stuff? I looked at the Transform.jl file and it seems the same as before.

galran commented:
Pushing it right now. Will probably finish in 3 min.

ran

From: BrianGun ***@***.***>
Sent: Friday, April 2, 2021 1:09 PM
To: microsoft/OpticSim.jl ***@***.***>
Cc: Ran Gal ***@***.***>; Mention ***@***.***>
Subject: Re: [microsoft/OpticSim.jl] Galran/new geometry module (#70)


@galran<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgalran&data=04%7C01%7Crgal%40microsoft.com%7C4c108e914ea44ed119ab08d8f6133086%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637529909576969943%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=ZNGfCy3Xhg%2F25vV2FfVm%2BnXvTGR%2BsZrfAK31Vzfp7lw%3D&reserved=0> did you push the latest stuff? I looked at the Transform.jl file and it seems the same as before.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2FOpticSim.jl%2Fpull%2F70%23issuecomment-812691167&data=04%7C01%7Crgal%40microsoft.com%7C4c108e914ea44ed119ab08d8f6133086%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637529909576969943%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=AOKKQYGJjYuSfBKqIyPrlb5ucfu4dEmnvV1EexL2z80%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FATBK6XSMIN23AFAULFW4OITTGYP6VANCNFSM42FULF2A&data=04%7C01%7Crgal%40microsoft.com%7C4c108e914ea44ed119ab08d8f6133086%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637529909576979941%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=%2Bh%2FivMcKlZpPUjdpKH%2FLqSv%2FIPA6dHXI91LYEjNQ6%2F8%3D&reserved=0>.
